src/mono/mono/metadata/appdomain.c

```diff
@@ -3261,6 +3262,7 @@ unload_thread_main (void *arg)
 
 	mono_loader_lock ();
 	mono_domain_lock (domain);
+	mono_memory_manager_lock (memory_manager);
 	/*
```

Comment:
This could cause problems since the memory manager lock is supposed to be a low level lock, and here it is taken for a large block of code which could take other locks etc.

Reply:
Short term, it might be a good idea to use the domain lock for locking the memory managers.